[QHC-831] Implement Square Waveform Optmization

fedonman opened the pull request:
We have introduced an optimization in the QbloxCompiler that significantly reduces memory usage when compiling square waveforms. The compiler now uses a heuristic algorithm that segments long waveforms into smaller chunks and loops over them. This optimization follows a two-pass search:

  1. **First Pass**: The compiler tries to find a chunk duration that divides the total waveform length evenly (i.e., remainder = 0).
  1. **Second Pass**: If no exact divisor is found, it looks for a chunk duration that leaves a remainder of at least 4 ns. This leftover chunk is large enough to be stored or handled separately.

  Each chunk duration is restricted to the range (\[100, 500\]) ns, ensuring that chunks are neither too small (leading to excessive repetitions) nor too large (risking out-of-memory issues). If no duration within (\[100, 500\]) ns meets these remainder constraints, the compiler defaults to using the original waveform in its entirety.

jjmartinezQT commented:
> _ No description provided. _

A description would be good :D altho this one I know

jjmartinezQT commented on src/qililab/qprogram/qblox_compiler.py:
```diff
@@ -591,19 +595,59 @@ def _handle_acquire(self, element: Acquire):
-        if not waveform_variables:
-            index_I, index_Q, duration = self._append_to_waveforms_of_bus(
+        if waveform_variables:
+            logger.error("Variables in waveforms are not supported in Qblox.")
```
nice one

jjmartinezQT commented on src/qililab/qprogram/qblox_compiler.py:
```diff
@@ -669,3 +713,30 @@ def _hash_waveform(waveform: Waveform):
+            return final_pd, duration // final_pd, duration % final_pd
+
+        # If no suitable piece_duration found, fallback to entire duration
+        return duration, 1, 0
```
Do you think is worthy raising a warning here telling the user it tried to optimise for values that do not allow optimisation?

fedonman replied:
not really, this is just a fallback e.g. for small waveforms like 40ns that will stay the same.

jjmartinezQT replied:
Cool, good on my side once changelog and codecoverage have been sorted. good job.

jjmartinezQT reviewed (approved):
Great improve. I approve it because it is up to the user to use this functionality and by default everything else will work as before, so we can allow this one to be merged before testing in hardware I think, but please test it as soon as we are back from xmas and fridges are cold.

fedonman commented:
> > _ No description provided. _
> 
> A description would be good :D altho this one I know

Yeah I will write the changelog and make some minor changes on Monday.

jjmartinezQT commented:
> > > _ No description provided. _
> > 
> > 
> > A description would be good :D altho this one I know
> 
> Yeah I will write the changelog and make some minor changes on Monday.

Cool, thanks!

jjmartinezQT reviewed (approved):
good one, make sure this does not fall forgotten in getting spread because of xmas :D We can start including this kind of changes in the tech newsletter

fedonman commented:
> good one, make sure this does not fall forgotten in getting spread because of xmas :D We can start including this kind of changes in the tech newsletter

Added changelog. For some reason the fallback return is not counted in codecov although I have added cases I have checked they do fallback. I will ignore it and merge it.
